README.md
Summary
=======

mp3cat copies a byte stream, skipping everything except mp3 frames with valid headers.

    $ mp3cat [options] - - < infile.mp3 > outfile.mp3
    $ mp3cat [options] - outdir < infile.mp3
    $ mp3cat [options] indir - > outfile.mp3
    $ mp3cat [options] indir outdir

Installation
============

    mkdir -p /usr/local/src
    cd /usr/local/src
    git clone https://github.com/tomclegg/mp3cat
    cd mp3cat
    make install

Filtering mp3 files
===================

mp3cat can read a stream from standard input and write to standard output:

    $ mp3cat - - < infile.mp3 > outfile.mp3

mp3cat only outputs MP3 frames with valid headers, even if there is extra garbage in its input stream.

Writing to an mp3dir
====================

Instead of standard output, mp3cat can store MP3 data across multiple files in a directory (an "mp3dir"). An mp3dir contains a number of files called tTIMESTAMP.mp3 (where TIMESTAMP is a UNIX timestamp) and one file called current.mp3. mp3cat appends MP3 frames to current.mp3. When current.mp3 exceeds a size threshold (default 1000000 bytes) while writing an mp3 frame, mp3cat renames it to tTIMESTAMP.mp3 (where TIMESTAMP is the current time) and starts a new current.mp3.

    $ mkdir /var/log/mp3files
    $ mp3cat - /var/log/mp3files < some-large-file.mp3
    $ ls /var/log/mp3files
    current.mp3
    t1033516598.mp3
    t1033516599.mp3

Control the size of mp3 segments with the --size parameter:

    $ mp3cat --size 14745600 - /var/log/mp3files < some-large-file.mp3

You can make an audio logger by reading from your sound card and writing to an mp3dir. Scripts to set this up are included. If svscan is running in /service, you can set up a logging service like this:

    # set up logging and cleaning services:
    cd /var/service
    pw useradd mp3log -d /nonexistent -s /usr/bin/true
    pw useradd mp3logl -d /nonexistent -s /usr/bin/true
    mp3log-conf mp3log mp3logl /var/service/mp3log /var/log/mp3log /dev/dsp 32

    # start the services
    chown root:mp3log /dev/dsp
    chmod 660 /dev/dsp
    ln -s /var/service/mp3log /service/
    ln -s /var/service/mp3log-clean /service/

Exceptions
==========

If tTIMESTAMP.mp3 already exists when the size threshold is reached, mp3cat keeps writing to current.mp3. Therefore, output files will be larger than expected if the size threshold is reached in less than one second.

If a write fails (perhaps due to a full disk), mp3cat waits a second and tries again.

Reading from an mp3dir
======================

Instead of standard input, mp3cat can also read MP3 data from an mp3dir. It reads from the end of current.mp3 just as tail -f would: when it reaches the end of current.mp3, it waits for more data to appear. While waiting for more data, if mp3cat detects that current.mp3 has been renamed and a new current.mp3 has been created, it closes the old file, opens the new one, and starts reading from the start of the new file.

You can turn your audio logger into an MP3 radio station by reading from the logger's mp3dir and writing to a network client. Here is a guide to using mp3cat to make an MP3 logger and MP3 streaming server.

Options
=======

The following options are accepted before indir and outdir arguments.

<dl>
<dt>--noclean</dt>
<dd>Output entire input data, even if it contains stuff other than mp3 frames.</dd>
<dt>--size bytes</dt>
<dd>Set target size for mp3 files (mp3dir output only; default 1000000)</dd>
<dt>--tail bytes</dt>
<dd>How far from the end of current.mp3 to start reading (mp3dir input only; default 16384)</dd>
<dt>-v</dt>
<dd>Print more status information to stderr</dd>
</dl>

Feedback
========

I'm happy to hear whether mp3cat works for you, and how you're using it.

License
=======

GPLv2. See COPYING file.


CHANGELOG
CHANGELOG for mp3cat

      + add mp3log-conf

      + write_buf.c: define O_NOFOLLOW if not defined by system header
	files (might help when compiling under Linux)

Release: 0.1


frame.c
/* $Id: frame.c,v 1.1.1.1 2003/05/04 22:02:23 tom Exp $ */

#include "frame.h"
#include <stdio.h>

const static int
v1l1_bitrate[] = {-1,32,64,96,128,160,192,224,256,288,320,352,384,416,448,-1};
const static int
v1l2_bitrate[] = {-1,32,48,56,64,80,96,112,128,160,192,224,256,320,384,-1};
const static int
v1l3_bitrate[] = {-1,32,40,48,56,64,80,96,112,128,160,192,224,256,320,-1};
const static int
v2l1_bitrate[] = {-1,32,48,56,64,80,96,112,128,144,160,176,192,224,256,-1};
const static int
v2l2or3_bitrate[] = {-1,8,16,24,32,40,48,56,64,80,96,112,128,144,160,-1};

const static int samplerate_table[4][3] =
{ {11025, 12000, 8000}, /* version 2.5 */
  {-1, -1, -1}, /* reserved version */
  {22050, 24000, 16000}, /* version 2 */
  {44100, 48000, 32000} /* version 1 */
};

int frame_length (const char* const buf, const int bufsize, int verbose, int *save_version, int *save_layer)
     /* returns expected length of this frame based on header,
	or 0 if header is incomprehensible */
{
  int
    version,
    layer,
    bitrate, bitrate_index,
    samplerate, samplerate_index,
    padding,
    length;

  if (bufsize < 4)
    return 0;

  /* make sure first 11 bits are on */
  if (buf[0] != '\377')
    return 0;
  if (buf[1] & '\340' != '\340')
    return 0;

  /* discover version: 0=2.5, 1=reserved, 2=2, 3=1  */
  #define VERSION_1 3
  #define VERSION_2 2
  #define VERSION_25 0
  version = (buf[1] >> 3) & 3;
  if (version == 1)
    return 0;

  /* discover layer: 0=reserved, 1=III, 2=II, 3=I */
  #define LAYER_I 3
  #define LAYER_II 2
  #define LAYER_III 1
  layer = (buf[1] >> 1) & 3;
  if (layer == 0)
    return 0;

  /* ignore protection bit: 0=protected, 1=not */

  /* discover bitrate */
  bitrate_index = (buf[2] >> 4) & 15;
  bitrate = 0;
  switch (version) {
  case VERSION_25:
  case VERSION_2:
    switch (layer) {
    case LAYER_III:
    case LAYER_II:
      bitrate = v2l2or3_bitrate[bitrate_index];
      break;
    case LAYER_I:
      bitrate = v2l1_bitrate[bitrate_index];
      break;
    default: /* should have already returned above! */
      return 0;
    }
    break;
  case VERSION_1:
    switch (layer) {
    case LAYER_III:
      bitrate = v1l3_bitrate[bitrate_index];
      break;
    case LAYER_II:
      bitrate = v1l2_bitrate[bitrate_index];
      break;
    case LAYER_I:
      bitrate = v1l1_bitrate[bitrate_index];
      break;
    default: /* should have already returned above! */
      return 0;
    }
    break;
  default: /* should have already returned above! */
    return 0;
  }
  /* some bitrates (0000, 1111) are invalid/undefined */
  if (bitrate <= 0)
    return 0;
  bitrate = bitrate * 1000;

  /* discover sample rate: 00, 01, 10, 11=reserved */
  samplerate_index = (buf[2] >> 2) & 3;
  if (samplerate_index == 3)
    return 0;
  samplerate = samplerate_table[version][samplerate_index];

  /* discover padding flag: 0=not padded, 1=padded */
  padding = (buf[2] >> 1) & 1;

  /* compute frame length: */
  if (layer == LAYER_I) {
    length = (12 * bitrate / samplerate + padding) * 4;
  } else {
    /* layer II and III */
    if (version == VERSION_1 || layer != LAYER_III)
      length = 144 * bitrate / samplerate + padding;
    else
      /* Judging by mpglib/common.c in lame source code -- and
         contrary to popular documentation -- MPEG-2 layer III and
         MPEG-2.5 layer III frames are half the size of other layer II
         and III frames: */
      length = 72 * bitrate / samplerate + padding;
  }

  if (verbose)
    fprintf (stderr, "version=%d layer=%d bitrate=%d samplerate=%d (%d) padding=%d length=%d\n",
	     version, layer, bitrate, samplerate, samplerate_index, padding, length);

  *save_version = version;
  *save_layer = layer;
  return length;
}


read_buf.c
/* $Id: read_buf.c,v 1.1.1.1 2003/05/04 22:02:23 tom Exp $ */

#include "read_buf.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <stdio.h>

static int Infile_is_open=0;
static int Infd=STDIN_FILENO;

int read_buf (char* buffer,
	      int wanttoread,
	      const char* const infile,
	      const int tail_bytes)
{
  int bytesread = 0, hasrolled, newfd, sysread_bytes;
  struct stat statbuf, fstatbuf;

  if (!Infile_is_open && infile) {
    while (0 > (newfd = open (infile, O_RDONLY))) {
      fprintf (stderr, "Couldn't open %s (%s), trying again\n",
	       infile,
	       strerror (errno));
      sleep (1);
    }
    lseek (newfd, -tail_bytes, SEEK_END);
    if (Infd != STDIN_FILENO)
      close (Infd);
    Infd = newfd;
    Infile_is_open = 1;
  }

  while (wanttoread > 0) {
    hasrolled =
      0 != infile
      &&
      0 == stat (infile, &statbuf)
      &&
      0 == fstat (Infd, &fstatbuf)
      &&
      statbuf.st_ino != fstatbuf.st_ino;
    sysread_bytes = read (Infd, buffer, wanttoread);
    if (infile == 0)
      return (sysread_bytes > 0 ? sysread_bytes : 0);
    if (sysread_bytes > 0) {
      buffer += sysread_bytes;
      wanttoread -= sysread_bytes;
      bytesread += sysread_bytes;
    }
    if (wanttoread > 0) {
      if (hasrolled) {
	if (Infd != STDIN_FILENO)
	  close (Infd);
	Infd = open (infile, O_RDONLY);
      }
      usleep (100);
    }
  }
  return bytesread;
}

#ifdef COMPILER_CAN_READ_PSEUDOCODE

read () {
  while (reading_dir && eof ()) {
    has_rolled
      = reading_dir && (current file's inode != current.mp3's inode);
    if (has_rolled && eof ()) {
      close();
      open("<indir>/current.mp3");
    } else {
      usleep(100);
    }
  }
  sysread ();
  return bytes_read;
}

#endif


write_buf.c
/* $Id: write_buf.c,v 1.1.1.1 2003/05/04 22:02:23 tom Exp $ */

#include "read_buf.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#ifndef O_NOFOLLOW
#define O_NOFOLLOW 0
#endif

static int Outfile_is_open=0;
static time_t Last_timestamp=0;
static int Outfd=STDOUT_FILENO;

int write_buf (const char* buffer, int wanttowrite,
	       const char* const outfile,
	       const char* const outfile_prefix,
	       const int bytes_per_file)
{
  int syswrite_bytes, newfd;
  struct stat statbuf;
  time_t current_time;
  char* newfilename;

  if (outfile && !Outfile_is_open) {
    while (0 > (newfd = open(outfile, O_CREAT|O_WRONLY|O_NOFOLLOW, 0x1ff))) {
      fprintf(stderr, "Couldn't open %s (%s), trying again\n", outfile, strerror(errno));
      sleep (1);
    }
    lseek (newfd, 0, SEEK_END);
    if (Outfd != STDOUT_FILENO)
      close (Outfd);
    Outfd = newfd;
  }
  if (outfile) {
    fstat (Outfd, &statbuf);
    if (statbuf.st_size >= bytes_per_file
	&& time(&current_time) > Last_timestamp) {
      Last_timestamp = current_time;
      while (!(newfilename = malloc (strlen(outfile_prefix)+21))) {
	fprintf(stderr, "Couldn't allocate %d bytes, trying again\n", strlen(outfile_prefix)+21);
	sleep (1);
      }
      strcpy (newfilename, outfile_prefix);
      if (newfilename [strlen (newfilename) - 1] != '/')
	strcat (newfilename, "/");
      sprintf (&newfilename[strlen(newfilename)], "t%d.mp3", current_time);
      rename (outfile, newfilename);
      free (newfilename);
      while (0 > (newfd = open (outfile, O_CREAT|O_WRONLY|O_NOFOLLOW, 0x1ff))) {
	fprintf(stderr, "Couldn't open %s (%s), trying again\n", outfile, strerror(errno));
	sleep (1);
      }
      close (Outfd);
      lseek (newfd, 0, SEEK_END);
      if (Outfd != STDOUT_FILENO)
	close (Outfd);
      Outfd = newfd;
    }
  }
  while (wanttowrite > 0) {
    syswrite_bytes = write (Outfd, buffer, wanttowrite);
    if (syswrite_bytes > 0) {
      wanttowrite -= syswrite_bytes;
      buffer += syswrite_bytes;
    }
    else if (!outfile && syswrite_bytes < 0 && errno != EAGAIN) {
      fprintf(stderr, "Write failed (%s), giving up.\n", strerror(errno));
      exit (1);
    }
    if (wanttowrite > 0) { /* disk full? */
      if (syswrite_bytes >= 0)
	fprintf(stderr, "Couldn't write %d bytes (only wrote %d), trying again\n", wanttowrite, syswrite_bytes);
      else
	fprintf(stderr, "Couldn't write %d bytes (%s), trying again\n", wanttowrite, strerror(errno));
      sleep (1);
    }
  }
}

#ifdef COMPILER_CAN_READ_PSEUDOCODE

write (wanttowrite_buf, wanttowrite_bytes) {
  int syswrite_bytes;
  if (writing_dir && !file_open_for_writing) {
    while (!open("<outdir>/current.mp3",
                 O_APPEND|O_CREAT|O_WRONLY|O_NOFOLLOW)) {
      sleep (1);
    }
  }
  if (writing_dir
      && Current_size >= Bytes_per_file
      && timestamp() > last_file_timestamp) {
    last_file_timestamp = timestamp();
    close();
    rename("current.mp3", "<last_file_timestamp>.mp3");
    Current_size = 0;
    open("current.mp3", "w");
  }
  while (wanttowrite_bytes > 0) {
    syswrite_bytes = syswrite ()) < wanttowrite_bytes;
    if(syswrite_bytes > 0) {
      wanttowrite_bytes -= syswrite_bytes;
      wanttowrite_buf += syswrite_bytes;
      Current_size += syswrite_bytes;
    }
    if (wanttowrite_bytes > 0) { // disk full
      sleep (1);
    }
  }
}

#endif


mp3cat.c
/* $Id: mp3cat.c,v 1.1.1.1 2003/05/04 22:02:23 tom Exp $ */

/* mp3cat
 *
 * Copyright 2002 Tom Clegg
 * tom@tomclegg.net -- 306 Carbonate St, Nelson BC V1L 4P2 Canada
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#define BUFMAX 8192 /* at least 4 times size of largest expected MP3 frame */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include "frame.h"
#include "read_buf.h"
#include "write_buf.h"

int main(int argc, char * const * argv)
{
  int bytesread;
  char buffer[BUFMAX];
  int bufend=0;
  int bufbegin=0;
  int framesize, nextframesize;
  int nbytes=0;
  int nframes=0;
  int timetostop=0;
  int layer, version, nextlayer, nextversion;
  char *Infile = 0, *Outfile = 0;
  const char * Outfile_prefix = 0;
  int Bytes_per_file=1000000;
  int Tail_bytes=BUFMAX*2;
  int Verbose=0;
  int Clean=1;
  int arg;

  for (arg=1; arg<argc && argv[arg][0] == '-' && argv[arg][1] != '\0'; arg++) {
    if (argv[arg][1] == 'v') {
      for (Verbose=1; argv[arg][Verbose] == 'v'; Verbose++);
      if (argv[arg][Verbose] != '\0')
	fprintf (stderr, "Ignoring garbage after 'v' characters in arg '%s'\n",
		 argv[arg]);
    }
    else if (!strcmp (argv[arg], "--clean"))
      Clean=1;
    else if (!strcmp (argv[arg], "--noclean"))
      Clean=0;
    else if (arg+1<argc && !strcmp (argv[arg], "--size")) {
      int param;
      if (sscanf (argv[arg+1], "%d", &param))
	Bytes_per_file = param;
      else
	fprintf (stderr, "Couldn't read --size parameter '%s'; ignoring.\n",
		 argv[arg+1]);
      arg++;
    }
    else if (arg+1<argc && !strcmp (argv[arg], "--tail")) {
      int param;
      if (sscanf (argv[arg+1], "%d", &param))
	Tail_bytes = param;
      else
	fprintf (stderr, "Couldn't read --tail parameter '%s'; ignoring.\n",
		 argv[arg+1]);
      arg++;
    }
    else
      fprintf (stderr, "Invalid argument '%s'; ignoring.\n", argv[arg]);
  }
  if (argc - arg != 2) {
    fprintf (stderr,
"Usage: %s [options] { indir | - } { outdir | - }\n"
"\n"
"Options:\n"
"  --noclean     Output entire input data, even if it contains stuff\n"
"                other than mp3 frames.\n"
"  --size bytes  Set target size for mp3 files (mp3dir output only;\n"
"                default 1000000).\n"
"  --tail bytes  How far from the end of current.mp3 to start reading\n"
"                (mp3dir input only; default 16384).\n"
"  -v            Print more status information to stderr.\n"
"\n"
"Examples:\n"
"  %s - - < dirty.mp3 > clean.mp3\n"
"  {generate MP3 stream} | %s --size 99000 - /tmp/mp3dir\n"
"  %s -v --tail 8000 /tmp/mp3dir - | {send to network client}\n"
"\n"
"WWW: http://tomclegg.net/mp3cat\n"
"\n",
	     argv[0], argv[0], argv[0], argv[0]);
    exit (1);
  }

  if (!strcmp (argv[arg], "-"))
    Infile = 0;
  else {
    if (!(Infile = malloc (strlen (argv[arg]) + 21))) {
      fprintf (stderr, "Couldn't allocate %d bytes; giving up.", strlen (argv[arg]) + 21);
      exit (1);
    }
    strcpy (Infile, argv[arg]);
    if (Infile [strlen (Infile) - 1] != '/')
      strcat (Infile, "/");
    strcat (Infile, "current.mp3");
  }

  ++arg;

  if (!strcmp (argv[arg], "-"))
    Outfile = 0;
  else {
    Outfile_prefix = argv[arg];
    if (!(Outfile = malloc (strlen (argv[arg]) + 21))) {
      fprintf (stderr, "Couldn't allocate %d bytes; giving up.", strlen (argv[arg]) + 21);
      exit (1);
    }
    strcpy (Outfile, argv[arg]);
    if (Outfile [strlen (Outfile)] != '/')
      strcat (Outfile, "/");
    strcat (Outfile, "current.mp3");
  }

  #define FLUSHFRAMES if (nframes) { if (Verbose >= 1) fprintf(stderr, "%d frames\n", nframes); nframes=0; }
  #define FLUSHBYTES if (nbytes) { if (Verbose >= 1) fprintf(stderr, "%d junkbytes\n", nbytes); nbytes=0; }

  do {
    /* assert: bufend >= bufbegin */

    /* shift buffer if bufbegin has passed BUFMAX/4 */
    if (bufbegin > BUFMAX/4) {
      bufend -= bufbegin;
      memmove (buffer, &buffer[bufbegin], bufend);
      bufbegin=0;
    }

    /* replenish buffer if bufend is under BUFMAX/2 */
    if (bufend <= BUFMAX/2 && !timetostop) {
      bytesread = read_buf (&buffer[bufend], BUFMAX-bufend,
			    Infile, Tail_bytes);
      if (bytesread > 0)
	bufend += bytesread;
      else if (!Infile)
	timetostop = 1;
    }

    framesize = frame_length (&buffer[bufbegin], bufend-bufbegin, 0, &version, &layer);
    if (framesize > 0 &&
	(nextframesize = frame_length (&buffer[bufbegin+framesize], bufend-bufbegin-framesize, 0, &nextversion, &nextlayer)) &&
	nextversion == version && nextlayer == layer) {

      /* valid MP3 frame */

      ++nframes;
      FLUSHBYTES;

      if (0)
	frame_length (&buffer[bufbegin], bufend-bufbegin, 1, &version, &layer);
      write_buf (&buffer[bufbegin], framesize, Outfile, Outfile_prefix, Bytes_per_file);
      bufbegin += framesize;
    }
    else {

      /* not a valid MP3 frame */

      ++nbytes;
      FLUSHFRAMES;

      if (Verbose >= 2 && framesize > 0) {
	fprintf (stderr, "Found: ");
	framesize = frame_length (&buffer[bufbegin], bufend-bufbegin, 1, &version, &layer);
	fprintf (stderr, "...but the next frame looks wrong");
	if (nextframesize)
	  fprintf (stderr, " (version %d, layer %d)", nextversion, nextlayer);
	fprintf (stderr, "\n");
      }
      bufbegin ++;
    }

  } while (bufbegin < bufend);

  FLUSHBYTES;
  FLUSHFRAMES;

  exit(0);
}
